### PR TITLE
Update proxy wasm cpp sdk to work with envoy 1.16.x

### PR DESCRIPTION
**Pull request by Sodman**
Tested against the latest Gloo 1.5.0-beta (master) updated to use an envoy based on this PR - https://github.com/solo-io/envoy-gloo/pull/83/files and confirmed it's working.

For code reviewing, _most_ of the changes in the diff are from moving the `examples` folder down a level into the `cli` package.

Things to look at when reviewing:
- New proxy-wasm-cpp-sdk added for Gloo 1.5.x (Latest ABI version 0_2_1). This is the latest used by upstream `envoy-wasm`.
- Added Istio 1.7 support, uses ABI version 0_2_0 so filters built for Istio 1.6 will not work with 1.7, and vice versa.
- Wasme will still support creating and deploying filters to both, but the filters themselves are not interchangeable as they were for Istio 1.5<->Istio1.6.

Fixes https://github.com/solo-io/wasm/issues/177
Fixes https://github.com/solo-io/wasm/issues/145

### CODE REVIEW

File: .github/workflows/pull_request.yaml [hunk 1]
```diff
@@ -106,8 +106,63 @@ jobs:
...
+        make install-deps run-tests
+    - name: Debug Info
+      if: failure()
+      env:
+        KUBECTL: ${{ steps.kubectl.outputs.kubectl-path }}
+      run: |
+        # see what's in the cluster if we failed
+        git diff
+        kubectl get all -A
+        kubectl get configmaps -A
```
Comment (ilackarms) [inline .github/workflows/pull_request.yaml:121]: nit: we should probably move this into a script for long term maintenance

Comment (ilackarms) [reply .github/workflows/pull_request.yaml:121]: or a make target

Comment (ilackarms) [reply .github/workflows/pull_request.yaml:121]: just a thought, i realize we have this elsewhere (and i'm not saying now is the ideal time to refactor)

Comment (Sodman) [reply .github/workflows/pull_request.yaml:121]: I agree, good call. Also in general I think we can do a bunch of cleanup on this GH Actions file to streamline it a bit. I'm thinking we can use the matrix approach we use elsewhere. I tried a few performance optimizations to get the build time down in this PR (as you might see from the commits) but most of them ended up maxing out the CPU of the kind cluster in the e2e test machines and ultimately being rolled back.

File: tools/wasme/cli/pkg/deploy/istio/istio_provider.go [hunk 2]
```diff
@@ -463,6 +486,22 @@ func (p *Provider) makeIstioEnvoyFilter(filter *v1.FilterSpec, image pull.Image,
...
 }
 
+// Returns true if istio version is 1.6.x or older
+func isOlderIstio(istioVersion string) bool {
+	parts := strings.Split(istioVersion, ".")
+
+	// check minor version
+	minor, err := strconv.Atoi(parts[1])
+	if err != nil {
+		logrus.WithField("istioVersion", istioVersion).Warn("unable to determine istio version, assuming 1.7+")
```
Comment (ilackarms) [inline tools/wasme/cli/pkg/deploy/istio/istio_provider.go:496]: might be helpful to include the error here as well

Comment (Sodman) [reply tools/wasme/cli/pkg/deploy/istio/istio_provider.go]: Added!

**Review by ilackarms — APPROVED** [review]

**Pull request merged**